src/typescript/decorators.ts

```diff
@@ -0,0 +1,125 @@
+ * property.
+ *
+ * This function must be invoked to return a decorator.
+ */
+export function property<T>(options?: PropertyOptions) {
+  return (proto: any, propName: string) : any => {
+    const notify : boolean = options && options.notify;
+    const type = Reflect.getMetadata("design:type", proto, propName);
+    const config = _ensureConfig(proto);
+    config.properties[propName] = {
+      type,
+      notify,
+    };
+  }
+}
+
+/**
+ * A TypeScript property decorator factory that causes the decorated method to
+ * be called when a property changes. `targets` is either a single property
+ * name, or a list of property names.
+ *
+ * This function must be invoked to return a decorator.
+ */
+export function observe(targets: string|string[]) {
+  return (proto: any, propName: string) : any => {
+    const targetString = typeof targets === 'string' ? targets : targets.join(',');
+    const config = _ensureConfig(proto);
+    config.observers.push(`${propName}(${targetString})`);
+  }
+}
+
+
+/**
+ * A TypeScript property decorator factory that converts a class property into a
+ * getter that executes a querySelector on the element's shadow root.
+ *
+ * By annotating the property with the correct type, element's can have
+ * type-checked access to internal elements.
+ *
+ * This function must be invoked to return a decorator.
+ */
+export const query = _query(
+    (target: NodeSelector, selector: string) => target.querySelector(selector));
+
+/**
+ * A TypeScript property decorator that converts a class property into a getter
+ * that executes a querySelectorAll on the element's shadow root.
+ *
+ * By annotating the property with the correct type, element's can have
+ * type-checked access to internal elements. The type should be NodeList
+ * with the correct type argument.
+ *
+ * This function must be invoked to return a decorator.
+ */
+export const queryAll = _query(
+    (target: NodeSelector, selector: string) => target.querySelectorAll(selector));
+
+
+interface Config {
+  properties: {[name: string]: PropertyDefinition};
+  observers: string[];
+}
+
+interface PropertyDefinition {
+  notify?: boolean;
+  type: Function;
+}
+
+function _ensureConfig(proto: any): Config {
+  const ctor = proto.constructor;
+  if (ctor.hasOwnProperty('__polymer_ts_config')) {
+    return ctor.__polymer_ts_config;
+  }
+
+  Object.defineProperty(ctor, 'config', {
+    get() { return ctor.__polymer_ts_config; }
+  });
+
+  const config: Config = ctor.__polymer_ts_config = ctor.__polymer_ts_config || {};
+  config.properties = config.properties || {};
+  config.observers = config.observers || [];
+  return config;
+}
+
+function _query(queryFn: (target: NodeSelector, selector: string) => Element|NodeList) {
+  return (selector: string) => (proto: any, propName: string): any => {
+    Object.defineProperty(proto, propName, {
+      get() {
+        return queryFn(this.shadowRoot, selector);
```

Comment:
```ts
return queryFn(this.shadowRoot, selector);                                                                       
                           ~~~~                                                                                              
                                                                                                                             
tmp/decorators.ts(116,24): error TS2683: 'this' implicitly has type 'any' because it does not have a type annotation.        
   
```